go/ql/lib/semmle/go/security/IncorrectIntegerConversionLib.qll

```diff
@@ -150,49 +151,169 @@ deprecated class ConversionWithoutBoundsCheckConfig extends TaintTracking::Confi
+ * data loss, and `isSigned` is true if the result is parsed as a signed
+ * integer.
+ */
+predicate isSourceWithBitSize(DataFlow::Node source, int bitSize, boolean isSigned) {
+  exists(DataFlow::CallNode c, IntegerParser::Range ip, int apparentBitSize |
+    c = ip.getACall() and
+    source = c.getResult(0) and
+    (
+      apparentBitSize = ip.getTargetBitSize()
+      or
+      // If we are reading a variable, check if it is
+      // `strconv.IntSize`, and use 0 if it is.
+      exists(DataFlow::Node rawBitSize |
+        rawBitSize = ip.getTargetBitSizeInput().getNode(c) and
+        if rawBitSize = any(Strconv::IntSize intSize).getARead()
+        then apparentBitSize = 0
+        else apparentBitSize = rawBitSize.getIntValue()
+      )
+    ) and
+    // Note that `getIntTypeBitSize` can return 0, so `effectiveBitSize`
+    // can be 0. Also `effectiveBitSize` is not necessarily the bit-size
+    // of an integer type - it can be any integer between 0 and 64.
+    bitSize = replaceZeroWith(apparentBitSize, getIntTypeBitSize(source.getFile(), 0)) and
+    isSigned = ip.isSigned()
+  )
+}
 
 private module ConversionWithoutBoundsCheckConfig implements DataFlow::StateConfigSig {
-  class FlowState = IntegerConversionFlowState;
+  class FlowState = MaxValueState;
 
   predicate isSource(DataFlow::Node source, FlowState state) {
-    exists(
-      DataFlow::CallNode c, IntegerParser::Range ip, int apparentBitSize, int effectiveBitSize
-    |
-      c.getTarget() = ip and source = c.getResult(0)
-    |
-      (
-        apparentBitSize = ip.getTargetBitSize()
-        or
-        // If we are reading a variable, check if it is
-        // `strconv.IntSize`, and use 0 if it is.
-        exists(DataFlow::Node rawBitSize | rawBitSize = ip.getTargetBitSizeInput().getNode(c) |
-          if rawBitSize = any(Strconv::IntSize intSize).getARead()
-          then apparentBitSize = 0
-          else apparentBitSize = rawBitSize.getIntValue()
+    exists(int effectiveBitSize, boolean sourceIsSigned |
+      isSourceWithBitSize(source, effectiveBitSize, sourceIsSigned) and
+      if effectiveBitSize = 0
+      then
+        exists(int b | b = [32, 64] |
+          state.getBitSize() = adjustBitSize(0, sourceIsSigned, b) and
+          state.getArchitectureBitSize() = b
         )
-      ) and
-      (
-        if apparentBitSize = 0
-        then effectiveBitSize = getIntTypeBitSize(source.getFile())
-        else effectiveBitSize = apparentBitSize
-      ) and
-      // `effectiveBitSize` could be any value between 0 and 64, but we
-      // can round it up to the nearest size of an integer type without
-      // changing behavior.
-      exists(int sourceBitSize |
-        sourceBitSize = min(int b | b in [0, 8, 16, 32, 64] and b >= effectiveBitSize)
-      |
-        state = TFlowstate(_, sourceBitSize, _)
+      else (
+        not exists(state.getArchitectureBitSize()) and
+        state.getBitSize() =
```

Comment:
add a comment on what this is doing -- assuming that the int could be 64-bit because the arch size is unknown I guess?

Reply:
This is the `else` branch for the conditional `if effectiveBitSize = 0`, so the value of that argument can never be used. I have added a comment to this effect.